pkg/reconciler/cache/publishedresources/publishedresources_controller.go

```diff
@@ -0,0 +1,298 @@
+	shardName string,
+	kcpClusterClient kcpclientset.ClusterInterface,
+	kcpCacheClient kcpclientset.ClusterInterface,
+	dynamicClient kcpdynamic.ClusterInterface,
+	cacheDynamicClient kcpdynamic.ClusterInterface,
+
+	kubeClusterClient kcpkubernetesclientset.ClusterInterface,
+	namespaceInformer kcpcorev1informers.NamespaceClusterInformer,
+	secretInformer kcpcorev1informers.SecretClusterInformer,
+
+	discoveringDynamicKcpInformers *informer.DiscoveringDynamicSharedInformerFactory,
+	cacheKcpInformers kcpinformers.SharedInformerFactory,
+
+	publishedResourceInformer cacheinformers.PublishedResourceClusterInformer,
+) (*Controller, error) {
+	c := &Controller{
+		shardName: shardName,
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.DefaultTypedControllerRateLimiter[string](),
+			workqueue.TypedRateLimitingQueueConfig[string]{
+				Name: ControllerName,
+			},
+		),
+		kcpClient:      kcpClusterClient,
+		kcpCacheClient: kcpCacheClient,
+
+		dynamicClient:      dynamicClient,
+		cacheDynamicClient: cacheDynamicClient,
+
+		discoveringDynamicKcpInformers: discoveringDynamicKcpInformers,
+		cacheKcpInformers:              cacheKcpInformers,
+
+		publishedResourceLister:  publishedResourceInformer.Lister(),
+		publishedResourceIndexer: publishedResourceInformer.Informer().GetIndexer(),
+		commit:                   committer.NewCommitter[*cachev1alpha1.PublishedResource, cachev1alpha1client.PublishedResourceInterface, *cachev1alpha1.PublishedResourceSpec, *cachev1alpha1.PublishedResourceStatus](kcpClusterClient.CacheV1alpha1().PublishedResources()),
+
+		secretNamespace: DefaultIdentitySecretNamespace,
+
+		getNamespace: func(clusterName logicalcluster.Name, name string) (*corev1.Namespace, error) {
+			return namespaceInformer.Lister().Cluster(clusterName).Get(name)
+		},
+		createNamespace: func(ctx context.Context, clusterName logicalcluster.Path, ns *corev1.Namespace) error {
+			_, err := kubeClusterClient.Cluster(clusterName).CoreV1().Namespaces().Create(ctx, ns, metav1.CreateOptions{})
+			return err
+		},
+		getSecret: func(ctx context.Context, clusterName logicalcluster.Name, ns, name string) (*corev1.Secret, error) {
+			secret, err := secretInformer.Lister().Cluster(clusterName).Secrets(ns).Get(name)
+			if err == nil {
+				return secret, nil
+			}
+
+			// In case the lister is slow to catch up, try a live read
+			secret, err = kubeClusterClient.Cluster(clusterName.Path()).CoreV1().Secrets(ns).Get(ctx, name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			return secret, nil
+		},
+		createSecret: func(ctx context.Context, clusterName logicalcluster.Path, secret *corev1.Secret) error {
+			_, err := kubeClusterClient.Cluster(clusterName).CoreV1().Secrets(secret.Namespace).Create(ctx, secret, metav1.CreateOptions{})
+			return err
+		},
+
+		conrollerRegistry: newRegistry(),
+	}
+
+	_, _ = publishedResourceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
The handler here will have to be deregistered on controller shutdown, otherwise goroutines are leaked on every deleted published resource.

Alternatively if `ClusterWithContext` is implemented that could be used here:

https://github.com/mjudeikis/kcp/blob/mjudeikis/cache.api/pkg/reconciler/cache/publishedresources/publishedresources_reconcile_replication.go#L62

I'll put in a comment there as well with the possible workaround

Reply:
I kinda was waiting for `ClusterWithContext` here :D Just that context will need to be created and stored together with controller yes :)

Reply:
So, for this particular place, I think we don't have issue (we have same as for every other controller). Like every other controller does this.